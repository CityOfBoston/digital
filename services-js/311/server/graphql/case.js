// @flow

import moment from 'moment-timezone';

import type { ServiceStub } from './service';
import type {
  ServiceRequest,
  DetailedServiceRequest,
} from '../services/Open311';
import type { IndexedCase } from '../services/Elasticsearch';

export type Root = ServiceRequest | DetailedServiceRequest | IndexedCase;

export const Schema = `
type Case {
  id: String!
  service: Service!
  description: String
  status: String!
  statusNotes: String
  address: String
  images: [CaseImage!]!
  location: LatLng
  requestedAt: Int!
  updatedAt: Int!
  requestedAtString(format: String): String!
  updatedAtString(format: String): String!
  requestedAtRelativeString: String!
  updatedAtRelativeString: String!
}

type CaseImage {
  tags: [String!]!
  squareThumbnailUrl: String!
  squarePreviewUrl: String!
  originalUrl: String!
}
`;

export type CaseImage = {
  tags: Array<string>,
  squareThumbnailUrl: string,
  squarePreviewUrl: string,
  originalUrl: string,
};

type DateStringArguments = {
  format?: string,
};

function makeHttpsImageUrl(mediaUrl: string): string {
  return (mediaUrl || '')
    .trim()
    .replace(
      'http://boston.spot.show/',
      'https://spot-boston-res.cloudinary.com/'
    );
}

function makeResizedImageUrls(url: string) {
  url = makeHttpsImageUrl(url);
  const imagePathMatch = url.match(/^(https?:\/\/.*\/image\/upload)\/(.*)$/);

  if (!imagePathMatch) {
    return {
      originalUrl: url,
      squarePreviewUrl: url,
      squareThumbnailUrl: url,
    };
  } else {
    return {
      originalUrl: url,
      squarePreviewUrl: [
        imagePathMatch[1],
        't_large_square_preview',
        imagePathMatch[2],
      ].join('/'),
      squareThumbnailUrl: [
        imagePathMatch[1],
        't_square_thumbnail',
        imagePathMatch[2],
      ].join('/'),
    };
  }
}

export const resolvers = {
  Case: {
    id: (r: Root) => r.service_request_id,
    service: (r: Root): ServiceStub => ({
      service_name: r.service_name || '',
      service_code: r.service_code || 'UNKNOWN',
    }),
    description: (r: Root) => r.description || '',
    status: (r: Root) => r.status,
    statusNotes: (r: Root) => r.status_notes || null,
    address: (r: Root) => r.address,
    images: (r: Root): Array<CaseImage> => {
      if (!r.media_url) {
        return [];
      } else if (Array.isArray(r.media_url)) {
        return r.media_url.map(i => ({
          tags: i.tags,
          ...makeResizedImageUrls(i.url),
        }));
      } else {
        return [
          {
            tags: [],
            ...makeResizedImageUrls(r.media_url),
          },
        ];
      }
    },
    location: (r: Root) => {
      if (r.lat != null && r.long != null) {
        return { lat: r.lat, lng: r.long };
      } else if (r.location) {
        return { lat: r.location.lat, lng: r.location.lon };
      } else {
        return null;
      }
    },
    requestedAt: (r: Root) => moment(r.requested_datetime).unix(),
    updatedAt: (r: Root) =>
      moment(r.updated_datetime || r.requested_datetime).unix(),
    // We format timezones on the server to avoid having to ship moment to the client
    requestedAtString: (r: Root, { format = '' }: DateStringArguments) =>
      moment(r.requested_datetime).tz('America/New_York').format(format),
    updatedAtString: (r: Root, { format = '' }: DateStringArguments) =>
      moment(r.updated_datetime || r.requested_datetime)
        .tz('America/New_York')
        .format(format),
    requestedAtRelativeString: (r: Root) =>
      moment(r.requested_datetime).fromNow(),
    updatedAtRelativeString: (r: Root) => moment(r.updated_datetime).fromNow(),
  },
};
