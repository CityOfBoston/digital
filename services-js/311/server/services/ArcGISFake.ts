import ArcGIS, { SearchResult, UnitResult } from './ArcGIS';

import SEARCH_RESULTS from './__fixtures__/arcgis/search.json';
import UNITS from './__fixtures__/arcgis/units.json';

export default class ArcGISFake implements Required<ArcGIS> {
  public async reverseGeocode(
    _lat: number,
    _lng: number
  ): Promise<SearchResult | null> {
    return null;
  }

  public async search(_query: string): Promise<SearchResult[]> {
    return SEARCH_RESULTS;
  }

  public async lookupUnits(_buildingId: string): Promise<UnitResult[]> {
    return UNITS;
  }
}
