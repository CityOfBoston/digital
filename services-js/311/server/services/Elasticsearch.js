// @flow

import AWS from 'aws-sdk';
import elasticsearch from 'elasticsearch';
import HttpAwsEs from 'http-aws-es';
import HttpsProxyAgent from 'https-proxy-agent';

// This needs to be kept up-to-date with the 311-indexer
type CaseDocument = {|
  status: string,
  location: ?{
    lat: number,
    lon: number,
  },
  address: string,
  description: string,
  service_name: string,
  service_code: ?string,
  status_notes: string,
  requested_datetime: ?string,
  updated_datetime: ?string,
  replay_id: number,
  media_url: ?string,
|};

export type IndexedCase = {
  service_request_id: string,
  ...CaseDocument,
};

type SearchHit<S> = {
  _index: string,
  _type: string,
  _id: string,
  _score: number,
  _source: S,
};

type SearchResponse<S> = {
  _shards: {
    total: number,
    successful: number,
    failed: number,
  },
  hits: {
    total: number,
    max_score: number,
    hits: Array<SearchHit<S>>,
    took: 5,
    timed_out: boolean,
  },
};

export default class Elasticsearch {
  opbeat: any;
  client: elasticsearch.Client;
  index: string;

  constructor(url: ?string, index: ?string, opbeat: any) {
    if (!url) {
      throw new Error('Missing Elasticsearch url');
    }

    if (!index) {
      throw new Error('Missing Elasticsearch index name');
    }

    this.client = new elasticsearch.Client({
      host: url,
      connectionClass: url.endsWith('.amazonaws.com') ? HttpAwsEs : undefined,
    });

    this.index = index;
    this.opbeat = opbeat;
  }

  static configureAws(region: ?string) {
    if (!region) {
      throw new Error('Missing AWS region');
    }

    const httpProxyUrl = process.env.http_proxy;
    const httpOptions = {
      agent: httpProxyUrl ? new HttpsProxyAgent(httpProxyUrl) : undefined,
    };

    AWS.config.update({
      region,
      httpOptions,
    });
  }

  // Returns an array of IDs for matching cases
  async searchCases(
    query: ?string,
    topLeft: ?{ lat: number, lng: number },
    bottomRight: ?{ lat: number, lng: number }
  ): Promise<Array<IndexedCase>> {
    const transaction =
      this.opbeat && this.opbeat.startTransaction('search', 'Elasticsearch');

    const must = [];
    const filter = [];

    if (query) {
      must.push({
        multi_match: {
          query: query || '',
          fields: ['description', 'service_name'],
        },
      });
    }

    if (topLeft && bottomRight) {
      filter.push({
        geo_bounding_box: {
          location: {
            top_left: {
              lat: topLeft.lat,
              lon: topLeft.lng,
            },
            bottom_right: {
              lat: bottomRight.lat,
              lon: bottomRight.lng,
            },
          },
        },
      });
    }

    try {
      const res: SearchResponse<CaseDocument> = await this.client.search({
        index: this.index,
        type: 'case',
        body: {
          size: 50,
          sort: [{ requested_datetime: 'desc' }],
          query: {
            bool: {
              must,
              filter,
            },
          },
        },
      });

      return res.hits.hits.map(h => ({
        service_request_id: h._id,
        ...h._source,
      }));
    } finally {
      if (transaction) {
        transaction.end();
      }
    }
  }
}
