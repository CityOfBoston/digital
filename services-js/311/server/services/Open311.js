// @flow

import 'isomorphic-fetch';
import URLSearchParams from 'url-search-params';
import url from 'url';
import HttpsProxyAgent from 'https-proxy-agent';
import DataLoader from 'dataloader';

// types taken from Open311
export type Service = {|
  service_code: string,
  service_name: ?string,
  description: string,
  metadata: boolean,
  type: 'realtime' | 'batch' | 'blackbox',
  keywords: string,
  group: string,
|}

export type PlainValue = {|
  key: string,
  name: string,
  img?: string,
|};

export type DependentCondition = {|
  attribute: string,
  op: 'eq' | 'neq' | 'in' | 'gt' | 'gte' | 'lt' | 'lte',
  // This is polymorphic; when comparing with a multipicklist this can 'eq'
  // an array of strings.
  value: number | string | string[],
|};

export type DependentConditions = {|
    clause: 'OR' | 'AND',
    conditions: DependentCondition[],
|};

export type ConditionalValues = {|
  dependentOn: DependentConditions,
  values: PlainValue[],
|};

export type ServiceMetadataAttributeValue = PlainValue | ConditionalValues;

export type ServiceMetadataAttribute = {|
  required: boolean,
  datatype: 'Text' | 'Informational' | 'Picklist' | 'Boolean (checkbox)',
  datatype_description: ?string,
  order: ?number,
  description: string,
  code: string,
  variable: boolean,
  values?: ServiceMetadataAttributeValue[],
  dependencies?: DependentConditions,
|};

export type ServiceMetadata = {|
  service_code: string,
  attributes: ServiceMetadataAttribute[],
  definitions: ?{|
    location_required?: boolean,
    contact_required?: boolean,
    location?: {|
      required: boolean,
      visible: boolean,
    |},
    reporter?: {|
      required: boolean,
      visible: boolean,
    |},
  |}
|}

export type ServiceRequest = {|
  service_request_id: string,
  status: string,
  status_notes: ?string,
  service_name: ?string,
  service_code: string,
  description: ?string,
  agency_responsible: ?string,
  service_notice: ?string,
  // 2017-02-21T22:18:27.000Z
  requested_datetime: string,
  updated_datetime: string,
  expected_datetime: ?string,
  address: ?string,
  address_id: ?string,
  zipcode: ?string,
  lat: ?number,
  long: ?number,
  media_url: ?string,
|};

export type CreateServiceRequestArgs = {|
  service_code: string,
  lat?: number,
  long?: number,
  address_string?: string,
  address_id?: string,
  first_name: ?string,
  last_name: ?string,
  email: ?string,
  phone: ?string,
  description: string,
  media_url: ?string,
  attributes: {code: string, value: string}[],
|};

async function processResponse(res): Promise<any> {
  if (res.status === 404) {
    return null;
  } else if (!res.ok) {
    let message;

    if (res.headers.get('content-type').startsWith('application/json')) {
      const firstError = (await res.json())[0];
      message = firstError.message || firstError.description || 'Open311 server error';
    } else {
      message = await res.text();
    }

    throw new Error(message);
  }

  return res.json();
}

/**
 * Service wrapper around our Open311 endpoint. Expected to be created fresh
 * for each request.
 *
 * Supports an HTTP proxy set via the $http_proxy env variable.
 *
 * Documentation: https://bos311.api-docs.io/
 */
export default class Open311 {
  agent: any;
  opbeat: any;
  endpoint: string;
  apiKey: ?string;
  serviceLoader: DataLoader<string, ?Service>;
  serviceMetadataLoader: DataLoader<string, ?ServiceMetadata>;
  requestLoader: DataLoader<string, ?ServiceRequest>;

  constructor(endpoint: ?string, apiKey: ?string, opbeat: any) {
    if (!endpoint) {
      throw new Error('Must specify an endpoint');
    }

    this.endpoint = endpoint;
    this.apiKey = apiKey;

    this.opbeat = opbeat;

    if (process.env.http_proxy) {
      this.agent = new HttpsProxyAgent(process.env.http_proxy);
    }

    // There's no API for just loading the name / code info for a single
    // service, so if we need that information we load all of the services
    // and then filter. We use DataLoader here to coallesce any simultaneous
    // requests for service lookup to a single API call.
    this.serviceLoader = new DataLoader(async (codes: string[]) => {
      const servicesByCode = {};
      (await this.services()).forEach((s) => { servicesByCode[s.service_code] = s; });
      return codes.map((code) => servicesByCode[code] || null);
    });

    this.serviceMetadataLoader = new DataLoader(async (codes: string[]) => {
      const transaction = opbeat && opbeat.startTransaction('serviceMetadata', 'Open311');
      const out = await Promise.all(codes.map(async (code) => {
        const params = new URLSearchParams();
        if (this.apiKey) {
          params.append('api_key', this.apiKey);
        }

        let additionalPath = '';
        if (process.env['311_METADATA_PATH']) {
          additionalPath = `${process.env['311_METADATA_PATH']}/`;
        }

        try {
          const response = await fetch(this.url(`services/${additionalPath}${code}.json?${params.toString()}`), {
            agent: this.agent,
          });

          return await processResponse(response);
        } catch (e) {
          throw new Error(`Error loading metadata for ${code}: ${e.toString()}`);
        }
      }));

      if (transaction) {
        transaction.end();
      }
      return out;
    });

    this.requestLoader = new DataLoader(async (ids: string[]) => {
      const transaction = opbeat && opbeat.startTransaction('request', 'Open311');
      const out = await Promise.all(ids.map(async (id) => {
        const params = new URLSearchParams();
        if (this.apiKey) {
          params.append('api_key', this.apiKey);
        }

        const response = await fetch(this.url(`requests/${id}.json?${params.toString()}`), {
          agent: this.agent,
        });

        // the endpoint returns the request in an array
        const requestArr: ?ServiceRequest[] = await processResponse(response);
        return requestArr ? requestArr[0] : null;
      }));

      if (transaction) {
        transaction.end();
      }
      return out;
    });
  }

  url(path: string) {
    return url.resolve(this.endpoint, path);
  }

  services = async (): Promise<Service[]> => {
    const transaction = this.opbeat && this.opbeat.startTransaction('services', 'Open311');
    const params = new URLSearchParams();
    if (this.apiKey) {
      params.append('api_key', this.apiKey);
    }

    const response = await fetch(this.url(`services.json?${params.toString()}`), {
      agent: this.agent,
    });

    const out = await processResponse(response) || [];
    if (transaction) {
      transaction.end();
    }
    return out;
  };

  service(code: string): Promise<?Service> {
    return this.serviceLoader.load(code);
  }

  serviceMetadata(code: string): Promise<?ServiceMetadata> {
    return this.serviceMetadataLoader.load(code);
  }

  request(id: string): Promise<?ServiceRequest> {
    return this.requestLoader.load(id);
  }

  requests = async () => {
    const transaction = this.opbeat && this.opbeat.startTransaction('requests', 'Open311');
    const params = new URLSearchParams();
    if (this.apiKey) {
      params.append('api_key', this.apiKey);
    }

    const response = await fetch(this.url(`requests.json?${params.toString()}`), {
      agent: this.agent,
    });

    const out = await processResponse(response) || [];
    if (transaction) {
      transaction.end();
    }
    return out;
  };


  async createRequest(args: CreateServiceRequestArgs): Promise<ServiceRequest> {
    const params = new URLSearchParams();
    if (this.apiKey) {
      params.append('api_key', this.apiKey);
    }

    Object.keys(args).forEach((key) => {
      switch (key) {
        case 'lat':
        case 'long':
          if (args[key] !== undefined) {
            params.append(key, args[key].toString());
          }
          break;
        case 'attributes':
          args[key].map(({ code, value }) => params.append(`attribute[${code}]`, value));
          break;
        default:
          params.append(key, args[key] || '');
      }
    });

    const body = params.toString();
    const bodyBuf = Buffer.from(body, 'utf-8');

    const response = await fetch(this.url('requests.json'), {
      agent: this.agent,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
        'Content-Length': `${bodyBuf.length}`,
      },
      method: 'POST',
      body,
    });

    return processResponse(response);
  }
}
