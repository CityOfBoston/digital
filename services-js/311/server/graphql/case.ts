import moment from 'moment-timezone';
import {
  ResolvableWith,
  Int,
  Resolvers,
} from '@cityofboston/graphql-typescript';

import { DetailedServiceRequest as Open311ServiceRequest } from '../services/Open311';
import { IndexedCase } from '../services/Elasticsearch';

import { Context } from '.';
import { ServiceStub, Service } from './service';
import { LatLng } from './query';

/**
 * We get a case either from Open311 or from our Elasticsearch index. They have
 * roughly the same data, so we want to re-use the resolvers between them.
 *
 * We call the field "request" because "case" is a reserved word.
 */
export type CaseRoot =
  | { source: 'Open311'; request: Open311ServiceRequest }
  | { source: 'Elasticsearch'; request: IndexedCase };

export interface Case extends ResolvableWith<CaseRoot> {
  id: string;
  service: Service;
  description: string | null;
  status: string;
  statusNotes: string | null;
  serviceNotice: string | null;
  closureReason: string | null;
  closureComment: string | null;
  address: string | null;
  images: CaseImage[];
  location: LatLng | null;
  requestedAt: Int | null;
  updatedAt: Int | null;
  expectedAt: Int | null;
  requestedAtString(args: { format?: string }): string | null;
  updatedAtString(args: { format?: string }): string | null;
  expectedAtString(args: { format?: string }): string | null;
  requestedAtRelativeString: string | null;
  updatedAtRelativeString: string | null;
  expectedAtRelativeString: string | null;
}

interface CaseImage {
  tags: string[];
  squareThumbnailUrl: string;
  squarePreviewUrl: string;
  originalUrl: string;
}

function makeHttpsImageUrl(mediaUrl: string): string {
  return (mediaUrl || '')
    .trim()
    .replace(
      'http://boston.spot.show/',
      'https://res.cloudinary.com/spot-boston/'
    )
    .replace(
      'http://spot-boston-res.cloudinary.com/',
      'https://spot-boston-res.cloudinary.com/'
    );
}

function makeResizedImageUrls(url: string) {
  url = makeHttpsImageUrl(url);
  const imagePathMatch = url.match(/^(https?:\/\/.*\/image\/upload)\/(.*)$/);

  if (!imagePathMatch) {
    return {
      originalUrl: url,
      squarePreviewUrl: url,
      squareThumbnailUrl: url,
    };
  } else {
    return {
      originalUrl: url,
      squarePreviewUrl: [
        imagePathMatch[1],
        't_large_square_preview',
        imagePathMatch[2],
      ].join('/'),
      squareThumbnailUrl: [
        imagePathMatch[1],
        't_square_thumbnail',
        imagePathMatch[2],
      ].join('/'),
    };
  }
}

function imagesFromMediaUrl(
  mediaUrl: Open311ServiceRequest['media_url'] | undefined
): CaseImage[] {
  if (!mediaUrl) {
    return [];
  } else if (Array.isArray(mediaUrl)) {
    return mediaUrl.map(i => ({
      tags: i.tags || [],
      ...makeResizedImageUrls(i.url),
    }));
  } else {
    return [
      {
        tags: [],
        ...makeResizedImageUrls(mediaUrl),
      },
    ];
  }
}

const caseResolvers: Resolvers<Case, Context> = {
  id: ({ request }) => request.service_request_id,
  service: ({ request }): ServiceStub => ({
    __type: 'ServiceStub',
    service_name: request.service_name || '',
    service_code: request.service_code || 'UNKNOWN',
  }),
  description: ({ request }) => request.description || '',
  status: ({ request }) => request.status,
  statusNotes: ({ request }) => request.status_notes || null,
  serviceNotice: caseRoot =>
    (caseRoot.source === 'Open311' && caseRoot.request.service_notice) || null,
  closureReason: caseRoot =>
    (caseRoot.source === 'Open311' &&
      caseRoot.request.closure_details.reason) ||
    null,
  closureComment: caseRoot =>
    (caseRoot.source === 'Open311' &&
      caseRoot.request.closure_details.comment) ||
    null,
  images: (caseRoot): CaseImage[] => {
    const caseImages = imagesFromMediaUrl(caseRoot.request.media_url);

    // Open311 cases may have images in the activities from being closed. The
    // Elasticsearch cases only have the media_url image.
    const activitiesImages =
      caseRoot.source === 'Open311'
        ? caseRoot.request.activities.map(({ media_url }) =>
            imagesFromMediaUrl(media_url)
          )
        : [];

    // Since the activities is an array of arrays, we use reduce to flatten
    // down to a single list of images.
    return [caseImages, ...activitiesImages].reduce(
      (arr, images) => [...arr, ...images],
      []
    );
  },
  address: ({ request }) => request.address || '',
  location: caseRoot => {
    switch (caseRoot.source) {
      case 'Elasticsearch':
        // Elasticsearch cases have the lat/lon in a "location" field because
        // that’s what’s required for use with Elasticsearch’s geo querying.
        return (
          caseRoot.request.location && {
            lat: caseRoot.request.location.lat,
            lng: caseRoot.request.location.lon,
          }
        );
      case 'Open311': {
        const { request } = caseRoot;

        if (
          request.reported_location &&
          request.reported_location.lat &&
          request.reported_location.long
        ) {
          // This is the lat/lng from the original submitter. It's preferred over
          // any other because it matches the submitter's intent.
          return {
            lat: request.reported_location.lat,
            lng: request.reported_location.long,
          };
        } else if (request.lat && request.long) {
          // Sometimes a lat/lng wasn't reported (e.g. just an address) so we use
          // the geocoded version from Salesforce.
          return { lat: request.lat, lng: request.long };
        } else {
          return null;
        }
      }
    }
  },
  requestedAt: ({ request }) =>
    request.requested_datetime
      ? moment(request.requested_datetime).unix()
      : null,
  updatedAt: ({ request }) => {
    // Some cases may not have an updated date, so fall back to the requested
    // one.
    const d = request.updated_datetime || request.requested_datetime;
    return d ? moment(d).unix() : null;
  },
  expectedAt: caseRoot =>
    caseRoot.source === 'Open311' && caseRoot.request.expected_datetime
      ? moment(caseRoot.request.expected_datetime).unix()
      : null,

  // We format timezones on the server to avoid having to ship moment to the client
  requestedAtString: async (caseRoot, { format = '' }, ctx) => {
    const requestedAt = await caseResolvers.requestedAt(caseRoot, {}, ctx);
    return requestedAt
      ? moment(requestedAt * 1000)
          .tz('America/New_York')
          .format(format)
      : null;
  },
  updatedAtString: async (caseRoot, { format = '' }, ctx) => {
    const updatedAt = await caseResolvers.updatedAt(caseRoot, {}, ctx);
    return updatedAt
      ? moment(updatedAt * 1000)
          .tz('America/New_York')
          .format(format)
      : null;
  },
  expectedAtString: async (caseRoot, { format = '' }, ctx) => {
    const expectedAt = await caseResolvers.expectedAt(caseRoot, {}, ctx);
    return expectedAt
      ? moment(expectedAt * 1000)
          .tz('America/New_York')
          .format(format)
      : null;
  },
  requestedAtRelativeString: async (caseRoot, _, ctx) => {
    const requestedAt = await caseResolvers.requestedAt(caseRoot, {}, ctx);
    return requestedAt ? moment(requestedAt * 1000).fromNow() : null;
  },
  updatedAtRelativeString: async (caseRoot, _, ctx) => {
    const updatedAt = await caseResolvers.updatedAt(caseRoot, {}, ctx);
    return updatedAt ? moment(updatedAt * 1000).fromNow() : null;
  },
  expectedAtRelativeString: async (caseRoot, _, ctx) => {
    const expectedAt = await caseResolvers.expectedAt(caseRoot, {}, ctx);
    return expectedAt ? moment(expectedAt * 1000).fromNow() : null;
  },
};

export const resolvers = {
  Case: caseResolvers,
};
