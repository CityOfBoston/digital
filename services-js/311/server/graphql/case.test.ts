import { resolvers, CaseRoot } from './case';
import { DetailedServiceRequest } from '../services/Open311';

const CASE_WITH_NO_PHOTOS: DetailedServiceRequest = {
  id: '500r0000002dwwsAAA',
  service_request_id: '18-00012429',
  status: 'closed',
  long: -71.05742059114559,
  lat: 42.358723671506326,
  media_url: null,
  service_name: 'Bicycle abandoned on the street or sidewalk',
  service_code: 'ABNDBIKE',
  description: 'bike',
  requested_datetime: '2018-02-02T18:09:34.000Z',
  expected_datetime: '2018-02-13T18:09:34.000Z',
  updated_datetime: '2018-02-02T18:09:34.000Z',
  address: '2 Devonshire Pl, Boston, 02109',
  zipcode: '02109',
  address_id: '164247',
  agency_responsible: null,
  service_notice: null,
  status_notes: null,
  origin: 'API',
  source: 'City Worker',
  priority: 'Low',
  service_level_agreement: {
    type: 7,
    value: 'Business Days',
  },

  address_details: 'detaila',
  duplicate_parent_service_request_id: null,
  parent_service_request_id: null,
  reported_location: {
    address: null,
    address_id: null,
    lat: 42.3587181,
    long: -71.0573287,
  },

  owner: {
    id: '00Gr0000000XANVEA4',
    name: 'BTDT BostonBikes',
    type: 'queue',
  },

  contact: {
    first_name: 'Test',
    last_name: 'Test',
    phone: null,
    email: 'test@boston.gov',
  },

  closure_details: {
    reason: 'Case Investigated',
    comment:
      'The bike was not found at this location when a Boston Transportation Department team member went to tag it. If the problem persists or if you have any questions, please feel free to contact us through BOS:311.',
  },

  attributes: [],
  activities: [],
  events: [],
};

const CASE_WITH_URL_PHOTO: CaseRoot = {
  source: 'Open311',
  request: {
    ...CASE_WITH_NO_PHOTOS,
    media_url:
      'https://spot-boston-res.cloudinary.com/image/upload/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg',
  },
};

const CASE_WITH_COMPLEX_PHOTO: CaseRoot = {
  source: 'Open311',
  request: {
    ...CASE_WITH_NO_PHOTOS,
    media_url: [
      {
        id: 'a0ir0000000DJLoAAO',
        url:
          'https://res.cloudinary.com/spot-boston/image/upload/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg',
        tags: ['Close'],
      },
    ],
  },
};

const CASE_WITH_ACTIVITY_PHOTO: CaseRoot = {
  source: 'Open311',
  request: {
    ...CASE_WITH_NO_PHOTOS,
    activities: [
      {
        code: 'ABNDBIKE-INSP2',
        order: 1,
        description: null,
        status: 'Complete',
        completion_date: '2018-03-21T16:56:30.000Z',
        media_url: [
          {
            id: 'a0ir0000000DJLoAAO',
            url:
              'https://res.cloudinary.com/spot-boston/image/upload/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg',
            tags: ['Close'],
          },
        ],
      },

      {
        code: 'ABNDBIKE-PICKUP',
        order: 2,
        description: null,
        status: 'Void',
        completion_date: null,
      },
    ],
  },
};

describe('photos', () => {
  it('returns a URL photo', () => {
    const images = resolvers.Case.images(CASE_WITH_URL_PHOTO, {}, {} as any);
    expect(images).toMatchInlineSnapshot(`
Array [
  Object {
    "originalUrl": "https://spot-boston-res.cloudinary.com/image/upload/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg",
    "squarePreviewUrl": "https://spot-boston-res.cloudinary.com/image/upload/t_large_square_preview/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg",
    "squareThumbnailUrl": "https://spot-boston-res.cloudinary.com/image/upload/t_square_thumbnail/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg",
    "tags": Array [],
  },
]
`);
  });

  it('returns a complex photo', () => {
    const images = resolvers.Case.images(
      CASE_WITH_COMPLEX_PHOTO,
      {},
      {} as any
    );
    expect(images).toMatchInlineSnapshot(`
Array [
  Object {
    "originalUrl": "https://res.cloudinary.com/spot-boston/image/upload/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squarePreviewUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_large_square_preview/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squareThumbnailUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_square_thumbnail/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "tags": Array [
      "Close",
    ],
  },
]
`);
  });

  it('gets photos from activities', () => {
    const images = resolvers.Case.images(
      CASE_WITH_ACTIVITY_PHOTO,
      {},
      {} as any
    );
    expect(images).toMatchInlineSnapshot(`
Array [
  Object {
    "originalUrl": "https://res.cloudinary.com/spot-boston/image/upload/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squarePreviewUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_large_square_preview/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squareThumbnailUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_square_thumbnail/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "tags": Array [
      "Close",
    ],
  },
]
`);
  });
});
