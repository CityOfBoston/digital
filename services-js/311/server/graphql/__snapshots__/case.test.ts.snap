// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`photos gets photos from activities 1`] = `
Array [
  Object {
    "originalUrl": "https://res.cloudinary.com/spot-boston/image/upload/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squarePreviewUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_large_square_preview/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squareThumbnailUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_square_thumbnail/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "tags": Array [
      "Close",
    ],
  },
]
`;

exports[`photos returns a URL photo 1`] = `
Array [
  Object {
    "originalUrl": "https://spot-boston-res.cloudinary.com/image/upload/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg",
    "squarePreviewUrl": "https://spot-boston-res.cloudinary.com/image/upload/t_large_square_preview/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg",
    "squareThumbnailUrl": "https://spot-boston-res.cloudinary.com/image/upload/t_square_thumbnail/v1521651380/boston/development/l79qpl0z81lmfzdt3rqd.jpg",
    "tags": Array [],
  },
]
`;

exports[`photos returns a complex photo 1`] = `
Array [
  Object {
    "originalUrl": "https://res.cloudinary.com/spot-boston/image/upload/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squarePreviewUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_large_square_preview/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "squareThumbnailUrl": "https://res.cloudinary.com/spot-boston/image/upload/t_square_thumbnail/v1493819948/boston/dev/ifonxqgpk8gukjkcvj78.jpg",
    "tags": Array [
      "Close",
    ],
  },
]
`;
